### PR TITLE
udev: priorize triggering of certain subsystems (e.g. block) over others

### PR DESCRIPTION
**Pull request by guozhijun111**
block device is critical path of bootup.
the activating time of xxx.device  is define as ueventtime - mount time.
but the block uevent is triggered whith all other type device.I think block uevent should trigger first.
the test result in my computer is(nvme0n1p3.device is the root device):
before modify:dev-nvme0n1p3.device 1140ms
after modify:dev-nvme0n1p3.device 720ms

### CODE REVIEW

File: units/systemd-udev-trigger.service [hunk 1]
```diff
@@ -19,5 +19,6 @@ ConditionPathIsReadWrite=/sys
 [Service]
 Type=oneshot
 RemainAfterExit=yes
-ExecStart=-udevadm trigger --type=subsystems --action=add
-ExecStart=-udevadm trigger --type=devices --action=add
+ExecStart=/bin/udevadm trigger --type=devices -s block --action=add
```
Comment (floppym) [inline units/systemd-udev-trigger.service]: There's no reason to replace `-udevadm` with `/bin/udevadm` here.

**Comment by floppym** [comment]
I wonder if there are any negative effects to scanning block devices before "subsystems".

**Comment by keszybz** [comment]
The comment on the pull request must be the message on the commit itself.

OK, I see how this *could* be faster. Do you have any measurements?

**Comment by yuwata** [comment]
No no no. This is not a way to go. systemd-udevd should process parents device first, then child. Even if this works on your system and improve boot performance, we cannot officially support such hack.

**Comment by guozhijun111** [comment]
>

**Comment by guozhijun111** [comment]
> 
In my systemd,the time to reach multi-user.target is:
before modify:2.71s
after modify:2.61s

**Comment by guozhijun111** [comment]
> No no no. This is not a way to go. systemd-udevd should process parents device first, then child. Even if this works on your system and improve boot performance, we cannot officially support such hack.

if systemd-udev-trigger.service is active,that means the root directory is mounted and block device is ready.

**Comment by boucman** [comment]
It does mean rootfs is up, but it does not always mean block devices are ready... only the block device where the rootfs is.

I mean... USB disk can literally appear at any point during boot.

**Comment by guozhijun111** [comment]
> It does mean rootfs is up, but it does not always mean block devices are ready... only the block device where the rootfs is.
> 
> I mean... USB disk can literally appear at any point during boot.

yes,you are right.so I think I can change it like this:

ExecStart=-udevadm trigger --type=devices -s block --action=add
ExecStart=-udevadm trigger --type=subsystems --action=add
ExecStart=-udevadm trigger --type=devices --action=add

if the system block device is all ready when systemd-udev-trigger active,the first line will trigger block uevent first.
if there is some block not ready,we can trigger block uevent  in the third line.in this case,there is no different with this commit

**Comment by poettering** [comment]
hmm, so this might be OKish if it's clear it has benefits. i mean, i can't deny that block devices are probably the primary type of devices we need to wait for, hence triggering them before the others might make sense. But I'd like some verification that this really has an effect.

Moreover, I'd really prefer if we'd add a new switch to "udevam trigger" called `--priorize-subsystem=` or so, that allows listing subsystems to trigger first. Would take a list, and we'd trigger those subsystems first, in the specified order, and then afterwards everything else. That way people can tweak the order on their specific devices if they need other devices than just block devices to be triggered first...

**Comment by poettering** [comment]
The priorization thing was discussed before, see: https://github.com/systemd/systemd/pull/19637#issuecomment-855966205

As back then I still think it would be the best way to go.

**Comment by yuwata** [comment]
A udev rule may have `ATTRS` or `TAGS` which references a attribute or tag of a parent device, and if a child block device is processed before its parent devices, then the result may be changed by this PR.

If a kind of reordering is applied (this PR or #19637), then we need to do more carefully to keep backward compatibility.

**Comment by poettering** [comment]
Yeah, it's fine to change the order of retriggering of leaves of the tree, but we should always retrigger from trunk to leaves, as first priority.

Thus an "udevadm trigger --priorize=" switch should probably iterate through all listed subsystems, go through all devices there, then first go up the tree trigger all parent devices, before triggering the enumerated devices themselves. It should keep track of all devices it triggered, so that it doesn't trigger them twice. Once it triggered all matching devices and their parents it should then do a sweep through the tree and trigger everything not triggered so far. 

That way we would guarantee that parents are always triggered and thus processed before their children, but the leaves are triggered in a bit smarter way.

**Comment by poettering** [comment]
I think #22662 is the right approach for this. Let's close this one here hence, and continue there.

**Pull request closed without merging**